src/SubscribedStream.js

```diff
@@ -17,10 +17,10 @@ export default class SubscribedStream {
     }
 
     async getProducers() {
-        if (!this.producers) {
-            this.producers = await this._client.getStreamProducers(this.streamId)
+        if (!this.producersPromise) {
+            this.producersPromise = this._client.getStreamProducers(this.streamId)
         }
-        return this.producers
+        return this.producersPromise
     }
 
     async verifyStreamMessage(msg) {
```

Comment:
So I believe the logic you want is the following (in case `client.options.verifySignatures == 'auto'`)?

- message signed AND stream requires signed data --> verify
- message signed AND stream doesn't require signed data --> verify
- message not signed AND stream requires signed data --> false
- message not signed AND stream doesn't require signed data --> true

I'm not sure about the second line, I am wondering if instead it should return 'true' and not verify the signature. Isn't it possible for a malicious node to add a random signature to received messages and forward it to other nodes who will reject it (kind of DoS attack)?
I am also wondering what would be a legitimate case where a signed message is produced on a stream that normally doesn't require signed data.

Reply:
Yes the table above is exactly what I meant.

> Isn't it possible for a malicious node to add a random signature to received messages and forward it to other nodes who will reject it (kind of DoS attack)?

Any malicious node could achieve the same effect by simply dropping the messages instead of going through the trouble of even spoofing a signature. In this case other (honest) nodes would still deliver the valid data. If the malicious node is the entry node, the client can notice that their messages are being censored and switch to another node, or produce to multiple nodes.

> I'm not sure about the second line, I am wondering if instead it should return 'true' and not verify the signature.

Hmm.. What reason would the client have *not* to verify a provided signature? It's extra information which can only benefit the client, right? A malicious node could remove the signature and tamper the data, but client's shouldn't trust unsigned data anyway, and this outcome is not worse than ignoring the signature.